src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        const copy = { ...currentState };
+        action.keys.forEach((key) => {
+          delete copy[key];
+        });
+
+        currentState = { ...copy };
+        break;
+
+      default:
+        break;
+    }
+
+    statesArray.push({ ...currentState });
```

Comment:
It's good that you're creating a new object to avoid mutating the previous state, but you can push the 'currentState' directly into 'statesArray' since you're already creating a new object in each case.